st2common/st2common/router.py

```diff
@@ -0,0 +1,386 @@
+# Licensed to the StackStorm, Inc ('StackStorm') under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import functools
+import re
+import six
+import sys
+import traceback
+
+import jsonschema
+from oslo_config import cfg
+import routes
+from six.moves.urllib import parse as urlparse  # pylint: disable=import-error
+from swagger_spec_validator.validator20 import validate_spec
+import webob
+from webob import exc, Request
+
+from st2common.exceptions import rbac as rbac_exc
+from st2common.exceptions import auth as auth_exc
+from st2common import log as logging
+from st2common.persistence.auth import User
+from st2common.rbac import resolvers
+from st2common.util.jsonify import json_encode
+from st2common.util.http import parse_content_type_header
+
+
+LOG = logging.getLogger(__name__)
+
+
+def op_resolver(op_id):
+    module_name, func_name = op_id.split(':', 1)
+    __import__(module_name)
+    module = sys.modules[module_name]
+    return functools.reduce(getattr, func_name.split('.'), module)
+
+
+def abort(status_code=exc.HTTPInternalServerError.code, message='Unhandled exception'):
+    raise exc.status_map[status_code](message)
+
+
+def abort_unauthorized(msg=None):
+    raise exc.HTTPUnauthorized('Unauthorized - %s' % msg if msg else 'Unauthorized')
+
+
+def extend_with_default(validator_class):
+    validate_properties = validator_class.VALIDATORS["properties"]
+
+    def set_defaults(validator, properties, instance, schema):
```

Comment:
IIRC, only of the problems with the default snippet is that it only handles validation for default values, but it doesn't actually manipulate the instance passed to the validator aka also assign the default values.

But yeah, if assigning is not needed, then it should work fine.